service/src/main/java/hu/uni/miskolc/iit/softwaretesting/serviceImpl/ReaderBookServiceImpl.java

```diff
@@ -0,0 +1,126 @@
+        return dao.getAvailableBooks();
+    }
+
+    @Override
+    public Collection<Book> getAvailableBooksByTitle(String title) throws EmptyFieldException, BookNotFoundException {
+
+        if (isEmptyField(title))
+            throw new EmptyFieldException("The given value should not be empty!");
+
+        Collection<Book> results = dao.getAvailableBooks();
+        Collection<Book> alma = dao.getBooksByTitle(title);
+
+        for (Book i : results) {
+            for (Book j : alma) {
+                if (i.getIsbn() != j.getIsbn()) {
+                    results.remove(j);
+                }
+            }
+        }
+
+        return results;
+    }
+
+    @Override
+    public Collection<Book> getBooksByYear(int year) throws BookNotFoundException, EmptyFieldException {
+
+        if (isEmptyField(String.valueOf(year)))
+            throw new EmptyFieldException("The field should not be empty!");
+
+        return dao.getBooksByYear(year);
+    }
+
+    @Override
+    public void requestBook(Book book, Reader reader) throws NoAvailableInstanceException {
+        try {
+            Collection<BookInstance> bookInstances = dao.getAvailableInstancesOfBook(book);
+            Calendar creationDate = Calendar.getInstance();
+            Calendar expirationDate = Calendar.getInstance();
+            expirationDate.setTime(creationDate.getTime());
+            expirationDate.add(Calendar.DATE, 30);
+            Borrowing borrowing = new Borrowing(createBorrowID(), reader, creationDate.getTime(), expirationDate.getTime(), BorrowStatus.REQUESTED, ((List<BookInstance>)bookInstances).get(0));
+            dao.createBorrowing(borrowing);
+        } catch (BookInstanceNotFound bookInstanceNotFound) {
+            bookInstanceNotFound.printStackTrace();
+            throw new NoAvailableInstanceException("There is no book to be borrowed.");
+
+        } catch (AlreadyExistingBorrowingException e) {
+            e.printStackTrace();
+        } catch (BookNotFoundException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public Collection<Borrowing> showBorrowings(Reader reader) throws NotExistingReaderException, NotExistingBorrowingException {
+        if (reader.equals(null))
+            throw new NotExistingReaderException("The given reader is null!");
+
+        if (dao.getBorrowingsOfReader(reader) == null || dao.getBorrowingsOfReader(reader).size() == 0)
+            throw new NotExistingBorrowingException("There is no borrowings for this user!");
+
+        return dao.getBorrowingsOfReader(reader);
+    }
+
+    @Override
+    public Collection<Book> getAllBooks() throws BookNotFoundException {
+        return dao.getAllBooks();
+    }
+
+
+    private boolean isEmptyField(String field) {
+        if (field.equalsIgnoreCase("") || field.equals(null))
+            return true;
+
+        return false;
+    }
+
+
+    private Long createBorrowID() {
+        return (new Random()).nextLong();
```

Comment:
When we have large amount of borrows, Random can generate a currently registered ID (so collision occurs). I think we have to generate ID in database (for example with [AUTOINCREMENT](https://www.tutorialspoint.com/sqlite/sqlite_using_autoincrement.htm) [sequence in OracleDB]).

So in line 84, initializing Borrow with null or 0 value ID is enough.